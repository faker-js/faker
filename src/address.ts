import type { Faker } from '.';
import type { Fake } from './fake';
import type { Helpers } from './helpers';

let f: Fake['fake'];

export class Address {
  readonly Helpers: Helpers;

  constructor(private readonly faker: Faker) {
    f = this.faker.fake;
    this.Helpers = this.faker.helpers;

    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Address.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }

    // TODO @Shinigami92 2022-01-13: Need better strategy
    // @ts-expect-error
    this.direction.schema = {
      description:
        'Generates a direction. Use optional useAbbr bool to return abbreviation',
      sampleResults: ['Northwest', 'South', 'SW', 'E'],
    };
    // @ts-expect-error
    this.cardinalDirection.schema = {
      description:
        'Generates a cardinal direction. Use optional useAbbr boolean to return abbreviation',
      sampleResults: ['North', 'South', 'E', 'W'],
    };
    // @ts-expect-error
    this.ordinalDirection.schema = {
      description:
        'Generates an ordinal direction. Use optional useAbbr boolean to return abbreviation',
      sampleResults: ['Northwest', 'Southeast', 'SW', 'NE'],
    };
  }

  /**
   * Generates random zipcode from format. If format is not specified, the
   * locale's zip format is used.
   *
   * @method faker.address.zipCode
   * @param format
   */
  zipCode(format?: string) {
    // if zip format is not specified, use the zip format defined for the locale
    if (typeof format === 'undefined') {
      const localeFormat = this.faker.definitions.address.postcode;
      if (typeof localeFormat === 'string') {
        format = localeFormat;
      } else {
        format = this.faker.random.arrayElement(localeFormat);
      }
    }
    return this.Helpers.replaceSymbols(format);
  }

  /**
   * Generates random zipcode from state abbreviation. If state abbreviation is
   * not specified, a random zip code is generated according to the locale's zip format.
   * Only works for locales with postcode_by_state definition. If a locale does not
   * have a postcode_by_state definition, a random zip code is generated according
   * to the locale's zip format.
   *
   * @method faker.address.zipCodeByState
   * @param state
   */
  zipCodeByState(state: string) {
    const zipRange = this.faker.definitions.address.postcode_by_state[state];
    if (zipRange) {
      return this.faker.datatype.number(zipRange);
    }
    return this.faker.address.zipCode();
  }

  /**
   * Generates a random localized city name. The format string can contain any
   * method provided by faker wrapped in `{{}}`, e.g. `{{name.firstName}}` in
   * order to build the city name.
   *
   * If no format string is provided one of the following is randomly used:
   *
   * * `{{address.cityPrefix}} {{name.firstName}}{{address.citySuffix}}`
   * * `{{address.cityPrefix}} {{name.firstName}}`
   * * `{{name.firstName}}{{address.citySuffix}}`
   * * `{{name.lastName}}{{address.citySuffix}}`
   * * `{{address.cityName}}` when city name is available
   *
   * @method faker.address.city
   * @param format
   */
  city(format?: string | number) {
    const formats = [
      '{{address.cityPrefix}} {{name.firstName}}{{address.citySuffix}}',
      '{{address.cityPrefix}} {{name.firstName}}',
      '{{name.firstName}}{{address.citySuffix}}',
      '{{name.lastName}}{{address.citySuffix}}',
    ];

    if (!format && this.faker.definitions.address.city_name) {
      formats.push('{{address.cityName}}');
    }

    if (typeof format !== 'number') {
      format = this.faker.datatype.number(formats.length - 1);
    }

    return f(formats[format]);
  }

  /**
   * Return a random localized city prefix
   *
   * @method faker.address.cityPrefix
   */
  cityPrefix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_prefix
    );
  }

  /**
   * Return a random localized city suffix
   *
   * @method faker.address.citySuffix
   */
  citySuffix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_suffix
    );
  }

  /**
   * Returns a random city name
   *
   * @method faker.address.cityName
   */
  cityName(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.city_name
    );
  }

  /**
   * Returns a random localized street name
   *
   * @method faker.address.streetName
   */
  streetName(): string {
    let result: string;
    let suffix = this.faker.address.streetSuffix();
    if (suffix !== '') {
      suffix = ' ' + suffix;
    }

    switch (this.faker.datatype.number(1)) {
      case 0:
        result = this.faker.name.lastName() + suffix;
        break;
      case 1:
        result = this.faker.name.firstName() + suffix;
        break;
    }
    return result;
  }

  //
  // TODO: change all these methods that accept a boolean to instead accept an options hash.
  //

  /**
   * Returns a random localized street address
   *
   * @method faker.address.streetAddress
   * @param useFullAddress
   */
  streetAddress(useFullAddress: boolean = false): string {
    let address = '';
    switch (this.faker.datatype.number(2)) {
      case 0:
        address =
          this.Helpers.replaceSymbolWithNumber('#####') +
          ' ' +
          this.faker.address.streetName();
        break;
      case 1:
        address =
          this.Helpers.replaceSymbolWithNumber('####') +
          ' ' +
          this.faker.address.streetName();
        break;
      case 2:
        address =
          this.Helpers.replaceSymbolWithNumber('###') +
          ' ' +
          this.faker.address.streetName();
        break;
    }
    return useFullAddress
      ? address + ' ' + this.faker.address.secondaryAddress()
      : address;
  }

  /**
   * streetSuffix
   *
   * @method faker.address.streetSuffix
   */
  streetSuffix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.street_suffix
    );
  }

  /**
   * streetPrefix
   *
   * @method faker.address.streetPrefix
   */
  streetPrefix(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.street_prefix
    );
  }

  /**
   * secondaryAddress
   *
   * @method faker.address.secondaryAddress
   */
  secondaryAddress() {
    return this.Helpers.replaceSymbolWithNumber(
      this.faker.random.arrayElement(['Apt. ###', 'Suite ###'])
    );
  }

  /**
   * county
   *
   * @method faker.address.county
   */
  county(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.county
    );
  }

  /**
   * country
   *
   * @method faker.address.country
   */
  country(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.country
    );
  }

  /**
   * countryCode
   *
   * @method faker.address.countryCode
   * @param alphaCode default alpha-2
   */
  countryCode(alphaCode: string = 'alpha-2'): string {
    if (alphaCode === 'alpha-2') {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.country_code
      );
    }

    if (alphaCode === 'alpha-3') {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.country_code_alpha_3
      );
    }

    return this.faker.random.arrayElement(
      this.faker.definitions.address.country_code
    );
  }

  /**
   * state
   *
   * @method faker.address.state
   * @param useAbbr
   */
  // TODO christopher 2022-01-13: useAbbr not in use
  state(useAbbr?: boolean): string {
    return this.faker.random.arrayElement(this.faker.definitions.address.state);
  }

  /**
   * stateAbbr
   *
   * @method faker.address.stateAbbr
   */
  stateAbbr(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.state_abbr
    );
  }

  /**
   * latitude
   *
   * @method faker.address.latitude
   * @param max default is 90
   * @param min default is -90
   * @param precision default is 4
   */
  latitude(max: number = 90, min: number = -90, precision: number = 4): string {
    return this.faker.datatype
      .number({
        max: max,
        min: min,
        precision: parseFloat((0.0).toPrecision(precision) + '1'),
      })
      .toFixed(precision);
  }

  /**
   * longitude
   *
   * @method faker.address.longitude
   * @param max default is 180
   * @param min default is -180
   * @param precision default is 4
   */
  longitude(
    max: number = 180,
    min: number = -180,
    precision: number = 4
  ): string {
    return this.faker.datatype
      .number({
        max: max,
        min: min,
        precision: parseFloat((0.0).toPrecision(precision) + '1'),
      })
      .toFixed(precision);
  }

  /**
   *  direction
   *
   * @method faker.address.direction
   * @param useAbbr return direction abbreviation. defaults to false
   */
  direction(useAbbr: boolean = false) {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr
    );
  }

  /**
   * cardinal direction
   *
   * @method faker.address.cardinalDirection
   * @param useAbbr return direction abbreviation. defaults to false
   */
  cardinalDirection(useAbbr: boolean = false): string {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction.slice(0, 4)
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr.slice(0, 4)
    );
  }

  /**
   * ordinal direction
   *
   * @method faker.address.ordinalDirection
   * @param useAbbr return direction abbreviation. defaults to false
   */
  ordinalDirection(useAbbr: boolean = false): string {
    if (!useAbbr) {
      return this.faker.random.arrayElement(
        this.faker.definitions.address.direction.slice(4, 8)
      );
    }
    return this.faker.random.arrayElement(
      this.faker.definitions.address.direction_abbr.slice(4, 8)
    );
  }

  nearbyGPSCoordinate(
    coordinate?: number[],
    radius?: number,
    isMetric?: boolean
  ): string[] {
    function randomFloat(min: number, max: number): number {
      return Math.random() * (max - min) + min;
    }
    function degreesToRadians(degrees: number): number {
      return degrees * (Math.PI / 180.0);
    }
    function radiansToDegrees(radians: number): number {
      return radians * (180.0 / Math.PI);
    }
    function kilometersToMiles(miles: number): number {
      return miles * 0.621371;
    }
    function coordinateWithOffset(
      coordinate: number[],
      bearing: number,
      distance: number,
      isMetric: boolean
    ): number[] {
      const R = 6378.137; // Radius of the Earth (http://nssdc.gsfc.nasa.gov/planetary/factsheet/earthfact.html)
      const d = isMetric ? distance : kilometersToMiles(distance); // Distance in km

      const lat1 = degreesToRadians(coordinate[0]); //Current lat point converted to radians
      const lon1 = degreesToRadians(coordinate[1]); //Current long point converted to radians

      const lat2 = Math.asin(
        Math.sin(lat1) * Math.cos(d / R) +
          Math.cos(lat1) * Math.sin(d / R) * Math.cos(bearing)
      );

      let lon2 =
        lon1 +
        Math.atan2(
          Math.sin(bearing) * Math.sin(d / R) * Math.cos(lat1),
          Math.cos(d / R) - Math.sin(lat1) * Math.sin(lat2)
        );

      // Keep longitude in range [-180, 180]
      if (lon2 > degreesToRadians(180)) {
        lon2 = lon2 - degreesToRadians(360);
      } else if (lon2 < degreesToRadians(-180)) {
        lon2 = lon2 + degreesToRadians(360);
      }

      return [radiansToDegrees(lat2), radiansToDegrees(lon2)];
    }

    // If there is no coordinate, the best we can do is return a random GPS coordinate.
    if (coordinate === undefined) {
      return [this.faker.address.latitude(), this.faker.address.longitude()];
    }
    radius ||= 10.0;
    isMetric ||= false;

    // TODO: implement either a gaussian/uniform distribution of points in circular region.
    // Possibly include param to function that allows user to choose between distributions.

    // This approach will likely result in a higher density of points near the center.
    const randomCoord = coordinateWithOffset(
      coordinate,
      degreesToRadians(Math.random() * 360.0),
      radius,
      isMetric
    );
    return [randomCoord[0].toFixed(4), randomCoord[1].toFixed(4)];
  }

  /**
   * Return a random time zone
   *
   * @method faker.address.timeZone
   */
  timeZone(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.address.time_zone
    );
  }
}
