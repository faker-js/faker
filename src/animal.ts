import type { Faker } from '.';

export class Animal {
  constructor(private readonly faker: Faker) {
    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Animal.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * Returns a random dog breed.
   *
   */
  dog(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.dog);
  }

  /**
   * Returns a random cat breed.
   *
   */
  cat(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cat);
  }

  /**
   * Returns a random snake species.
   *
   */
  snake(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.snake);
  }

  /**
   * Returns a random bear species.
   *
   */
  bear(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bear);
  }

  /**
   * Returns a random lion species.
   *
   */
  lion(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.lion);
  }

  /**
   * Returns a random cetacean species.
   *
   * @method faker.animal.cetacean
   */
  cetacean(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.cetacean
    );
  }

  /**
   * Returns a random horse breed.
   *
   */
  horse(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.horse);
  }

  /**
   * Returns a random bird species.
   *
   */
  bird(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.bird);
  }

  /**
   * Returns a random cow species.
   *
   */
  cow(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.cow);
  }

  /**
   * Returns a random fish species.
   *
   */
  fish(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.fish);
  }

  /**
   * Returns a random crocodilian species.
   *
   */
  crocodilia(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.animal.crocodilia
    );
  }

  /**
   * Returns a random insect species.
   *
   */
  insect(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.insect);
  }

  /**
   * Returns a random rabbit species
   *
   */
  rabbit(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.rabbit);
  }

  /**
   * Returns a random animal type.
   *
   */
  type(): string {
    return this.faker.random.arrayElement(this.faker.definitions.animal.type);
  }
}
