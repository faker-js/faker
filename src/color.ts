import type { Faker } from '.';

type ColorSpace =
  | 'rgb'
  | 'rgba'
  | 'hsl'
  | 'hsla'
  | 'hwb'
  | 'cmyk'
  | 'lab'
  | 'lch'
  | 'display-p3';

/**
 * Formats the hex format of a generated color string according
 * to options specified by user.
 *
 * @param hexColor Hex color string to be formated.
 * @param options options object.
 * @param options.prefix Prefix of the generated hex color. Defaults to `0x`.
 * @param options.case Letter case of the generated hex color. Defaults to `mixed`.
 */
function applyHexFormat(
  hexColor: string,
  options?: {
    prefix?: string;
    case?: 'upper' | 'lower' | 'mixed';
  }
): string {
  if (options?.prefix) hexColor = hexColor.replace('0x', options.prefix);
  switch (options?.case) {
    case 'upper':
      hexColor = hexColor.toUpperCase();
      break;
    case 'lower':
      hexColor = hexColor.toLowerCase();
      break;
  }
  return hexColor;
}

/**
 * Converts an array of numbers into binary string format.
 *
 * @param values Array of values to be converted.
 */
function toBinary(values: number[]): string {
  const binary: string[] = values.map((value: number) => {
    const isFloat: boolean = Number(value) === value && value % 1 !== 0;
    if (isFloat) {
      const buffer: ArrayBuffer = new ArrayBuffer(4);
      new DataView(buffer).setFloat32(0, value);
      const bytes = new Uint8Array(buffer);
      return toBinary(Array.from(bytes)).split(' ').join('');
    }
    return (value >>> 0).toString(2).padStart(8, '0');
  });
  return binary.join(' ');
}

/**
 * Converts an array of numbers into CSS accepted format.
 *
 * @param values Array of values to be converted.
 * @param colorSpace Color space to format CSS string for. If invalid color space
 * is provided, RGB CSS will be returned.
 */
function toCSS(values: number[], colorSpace: ColorSpace): string {
  const percentages = values.map((value: number) => Math.round(value * 100));
  switch (colorSpace) {
    case 'rgb':
      return `rgb(${values[0]}, ${values[1]}, ${values[2]})`;
    case 'rgba':
      return `rgba(${values[0]}, ${values[1]}, ${values[2]}, ${values[3]})`;
    case 'cmyk':
      return `cmyk(${percentages[0]}%, ${percentages[1]}%, ${percentages[2]}%, ${percentages[3]}%)`;
    case 'hsl':
      return `hsl(${values[0]}deg ${percentages[1]}% ${percentages[2]}%)`;
    case 'hsla':
      return `hsl(${values[0]}deg ${percentages[1]}% ${percentages[2]}% / ${percentages[3]})`;
    case 'hwb':
      return `hwb(${values[0]} ${percentages[1]}% ${percentages[2]}%)`;
    case 'lab':
      return `lab(${percentages[0]}% ${values[1]} ${values[2]})`;
    case 'lch':
      return `lch(${percentages[0]}% ${values[1]} ${values[2]})`;
    case 'display-p3':
      return `color(display-p3 ${values[0]} ${values[1]} ${values[2]})`;
    default:
      return toCSS(values, 'rgb');
  }
}

/**
 * Converts an array of color values to the specified color format.
 *
 * @param values Array of color values to be converted.
 * @param format Format of generated RGB color.
 * @param colorSpace Color space to format CSS string for. Defaults to `rgb`.
 */
function toColorFormat(
  values: number[],
  format: 'decimal' | 'css' | 'binary',
  colorSpace?: ColorSpace
): string | number[] {
  colorSpace = colorSpace || 'rgb';
  if (format === 'decimal') return values;

  let result: string | number[];
  switch (format) {
    case 'css':
      result = toCSS(values, colorSpace);
      break;
    case 'binary':
      result = toBinary(values);
      break;
  }
  return result;
}

/**
 * Module to generate colors.
 */
export class Color {
  constructor(private readonly faker: Faker) {
    // Bind `this` so namespaced is working correctly
    for (const name of Object.getOwnPropertyNames(Color.prototype)) {
      if (name === 'constructor' || typeof this[name] !== 'function') {
        continue;
      }
      this[name] = this[name].bind(this);
    }
  }

  /**
   * Returns a random human readable color name.
   *
   * @example
   * faker.color.human() // 'red'
   */
  human(): string {
    return this.faker.random.arrayElement(this.faker.definitions.color.human);
  }

  /**
   * Returns a random color gamut name.
   *
   * @example
   * faker.color.colorGamut() // 'sRGB'
   */
  colorGamut(): string {
    return this.faker.random.arrayElement(
      this.faker.definitions.color.colorGamut
    );
  }

  /**
   * Returns a RGB color.
   *
   * @param options options object.
   * @param options.prefix Prefix of the generated hex color. Defaults to `0x`. Only applied when 'hex' format is used.
   * @param options.case Letter case of the generated hex color. Defaults to `mixed`. Only applied when 'hex' format is used.
   * @param options.format Format of generated RGB color. Defaults to `hex`.
   * @param options.includeAlpha Adds an alpha value to the color (RGBA). Defaults to `false`.
   *
   * @example
   * faker.color.rgb() // '0xffffFF'
   * faker.color.rgb({ prefix: '#' }) // '#ffffFF'
   * faker.color.rgb({ case: 'upper' }) // '0xFFFFFF'
   * faker.color.rgb({ case: 'lower' }) // '0xffffff'
   * faker.color.rgb({ prefix: '#', case: 'lower' }) // '#ffffff'
   * faker.color.rgb({ format: 'decimal' }) // '[255, 255, 255]'
   * faker.color.rgb({ format: 'css' }) // 'rgb(255, 0, 0)'
   * faker.color.rgb({ format: 'binary' }) // '10000000 00000000 11111111'
   * faker.color.rgb({ format: 'decimal', includeAlpha: true }) // '[255, 255, 255, 0.4]'
   */
  rgb(options?: {
    prefix?: string;
    case?: 'upper' | 'lower' | 'mixed';
    format?: 'hex' | 'decimal' | 'css' | 'binary';
    includeAlpha?: boolean;
  }): string | number[] {
    let color: string | number[];
    let colorSpace: ColorSpace = 'rgb';
    if (!options?.format) options = { ...options, format: 'hex' };
    if (options?.format === 'hex') {
      color = this.faker.datatype.hexadecimal(options?.includeAlpha ? 8 : 6);
      color = applyHexFormat(color, options);
      return color;
    }

    color = [0, 0, 0].map(() =>
      this.faker.datatype.number({ min: 0, max: 255 })
    );
    if (options?.includeAlpha) {
      color.push(this.faker.commerce.percentage(0.01));
      colorSpace = 'rgba';
    }
    return toColorFormat(color, options.format, colorSpace);
  }

  /**
   * Returns a CMYK color.
   *
   * @param options options object.
   * @param options.format Format of generated CMYK color. Defaults to `decimal`.
   *
   * @example
   * faker.color.cmyk() // [0.31, 0.52, 0.32, 0.43]
   * faker.color.cmyk({ format: 'decimal' }) // [0.31, 0.52, 0.32, 0.43]
   * faker.color.cmyk({ format: 'css' }) // cmyk(100%, 0%, 0%, 0%)
   * faker.color.cmyk({ format: 'binary' }) // (8-32 bits) x 4
   */
  cmyk(options?: { format?: 'decimal' | 'css' | 'binary' }): string | number[] {
    const color: string | number[] = [0, 0, 0, 0].map(() =>
      this.faker.commerce.percentage(0.01)
    );

    return toColorFormat(color, options?.format || 'decimal', 'cmyk');
  }

  /**
   * Returns a HSL color.
   *
   * @param options options object.
   * @param options.format Format of generated HSL color. Defaults to `decimal`.
   * @param options.includeAlpha Adds an alpha value to the color (RGBA). Defaults to `false`.
   *
   * @example
   * faker.color.hsl() // [201, 0.23, 0.32]
   * faker.color.hsl({ format: 'decimal' }) // [300, 0.21, 0.52]
   * faker.color.hsl({ format: 'decimal', includeAlpha: true }) // [300, 0.21, 0.52, 0.28]
   * faker.color.hsl({ format: 'css' }) // hsl(0deg, 100%, 80%)
   * faker.color.hsl({ format: 'css', includeAlpha: true }) // hsl(0deg 100% 50% / 0.5)
   * faker.color.hsl({ format: 'binary' }) // (8-32 bits) x 3
   * faker.color.hsl({ format: 'binary', includeAlpha: true }) // (8-32 bits) x 4
   */
  hsl(options?: {
    format?: 'decimal' | 'css' | 'binary';
    includeAlpha?: boolean;
  }): string | number[] {
    const hsl: number[] = [this.faker.datatype.number({ min: 0, max: 360 })];
    for (let i = 0; i < (options?.includeAlpha ? 3 : 2); i++) {
      hsl.push(this.faker.commerce.percentage(0.01));
    }
    return toColorFormat(
      hsl,
      options?.format || 'decimal',
      options?.includeAlpha ? 'hsla' : 'hsl'
    );
  }

  /**
   * Returns a HWB color.
   *
   * @param options options object.
   * @param options.format Format of generated RGB color. Defaults to `decimal`.
   *
   * @example
   * faker.color.hwb() // [201, 0.21, 0.31]
   * faker.color.hwb({ format: 'decimal' }) // [201, 0.21, 0.31]
   * faker.color.hwb({ format: 'css' }) // hwb(194 0% 0%)
   * faker.color.hwb({ format: 'binary' }) // (8-32 bits x 3)
   */
  hwb(options?: { format?: 'decimal' | 'css' | 'binary' }): string | number[] {
    const hsl: number[] = [this.faker.datatype.number({ min: 0, max: 360 })];
    for (let i = 0; i < 2; i++) {
      hsl.push(this.faker.commerce.percentage(0.01));
    }
    return toColorFormat(hsl, options?.format || 'decimal', 'hwb');
  }

  /**
   * Returns a LAB (CIELAB) color.
   *
   * @param options options object.
   * @param options.format Format of generated RGB color. Defaults to `decimal`.
   *
   * @example
   * faker.color.lab() // [0.832133, -80.3245, 100.1234]
   * faker.color.lab({ format: 'decimal' }) // [0.856773, -80.2345, 100.2341]
   * faker.color.lab({ format: 'css' }) // lab(29.2345% 39.3825 20.0664)
   * faker.color.lab({ format: 'binary' }) // (8-32 bits x 3)
   */
  lab(options?: { format?: 'decimal' | 'css' | 'binary' }): string | number[] {
    const lab = [this.faker.commerce.percentage(0.000001)];
    for (let i = 0; i < 2; i++) {
      lab.push(
        this.faker.datatype.float({ min: -100, max: 100, precision: 0.0001 })
      );
    }
    return toColorFormat(lab, options?.format || 'decimal', 'lab');
  }

  /**
   * Returns a LCH color. Even though upper bound of
   * chroma in LCH color space is theoretically unbounded,
   * it is bounded to 230 as anything above will not
   * make a noticable difference in the browser.
   *
   * @param options options object.
   * @param options.format Format of generated RGB color. Defaults to `decimal`.
   *
   * @example
   * faker.color.lch() // [0.522345, 72.2, 56.2]
   * faker.color.lch{ format: 'decimal' }) // [0.522345, 72.2, 56.2]
   * faker.color.lch{ format: 'css' }) // lch(52.2345% 72.2 56.2)
   * faker.color.lch{ format: 'binary' }) // (8-32 bits x 3)
   */
  lch(options?: { format?: 'decimal' | 'css' | 'binary' }): string | number[] {
    const lch = [this.faker.commerce.percentage(0.000001)];
    for (let i = 0; i < 2; i++) {
      lch.push(
        this.faker.datatype.number({ min: 0, max: 230, precision: 0.1 })
      );
    }
    return toColorFormat(lch, options?.format || 'decimal', 'lch');
  }

  /**
   * Return a display-p3 color.
   *
   * @param options options object.
   * @param options.format Format of generated RGB color. Defaults to `decimal`.
   *
   * @example
   * faker.color.displayP3() // [0.93, 1, 0.82]
   * faker.color.displayP3({ format: 'decimal' }) // [0.12, 0.21, 0.31]
   * faker.color.displayP3({ format: 'css' }) // color(display-p3 0.12 1 0.23)
   * faker.color.displayP3({ format: 'binary' }) // (8-32 bits x 3)
   */
  displayP3(options?: {
    format?: 'decimal' | 'css' | 'binary';
  }): string | number[] {
    const p3 = [0, 0, 0].map(() => this.faker.commerce.percentage(0.01));
    return toColorFormat(p3, options?.format || 'decimal', 'display-p3');
  }
}
