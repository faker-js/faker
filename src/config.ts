/**
 * The possible configuration options, that can be set.
 * This type exists to be extended for plugins via type augmentation.
 *
 * The `@default` tag is used to indicate the default value, that should be used if, the config is absent.
 */
export interface FakerConfig {
  /**
   * The function used to generate the `refDate` date instance, if not provided as method param.
   * The function must return a new valid `Date` instance for every call.
   *
   * @see [Reproducible Results](https://fakerjs.dev/guide/usage.html#reproducible-results)
   * @see faker.seed(): For generating reproducible values.
   *
   * @since 9.0.0
   *
   * @default () => new Date()
   */
  defaultRefDate?: () => Date;
}
