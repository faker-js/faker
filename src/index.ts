import { Faker } from './faker';
import allLocales from './locales';

export type {
  AnimalDefinitions,
  ColorDefinitions,
  CommerceDefinitions,
  CommerceProductNameDefinitions,
  CompanyDefinitions,
  DatabaseDefinitions,
  DateDefinitions,
  DateEntryDefinition,
  FinanceCurrencyEntryDefinitions,
  FinanceDefinitions,
  HackerDefinitions,
  InternetDefinitions,
  LocaleDefinition,
  /** @deprecated Use LocationDefinitions instead */
  LocationDefinitions as AddressDefinitions,
  LocationDefinitions,
  LoremDefinitions,
  MusicDefinitions,
  /** @deprecated Use PersonDefinitions instead */
  PersonDefinitions as NameDefinitions,
  PersonDefinitions,
  /** @deprecated Use PersonTitleDefinitions instead */
  PersonTitleDefinitions as NameTitleDefinitions,
  PersonTitleDefinitions,
  PhoneNumberDefinitions,
  ScienceDefinitions,
  SystemDefinitions,
  SystemMimeTypeEntryDefinitions,
  VehicleDefinitions,
  WordDefinitions,
} from './definitions';
export { FakerError } from './errors/faker-error';
export type { FakerOptions, UsableLocale, UsedLocales } from './faker';
export type { AnimalModule } from './modules/animal';
export type {
  Casing,
  ColorFormat,
  ColorModule,
  CSSFunction,
  CSSSpace,
  NumberColorFormat,
  StringColorFormat,
} from './modules/color';
export type { CommerceModule } from './modules/commerce';
export type { CompanyModule } from './modules/company';
export type { DatabaseModule } from './modules/database';
export type { DatatypeModule } from './modules/datatype';
export type { DateModule } from './modules/date';
export type { FinanceModule } from './modules/finance';
export type { GitModule } from './modules/git';
export type { HackerModule } from './modules/hacker';
export type { HelpersModule } from './modules/helpers';
export type { ImageModule } from './modules/image';
export type { InternetModule } from './modules/internet';
export type {
  /** @deprecated Use LocationModule instead */
  LocationModule as AddressModule,
  LocationModule,
} from './modules/location';
export type { LoremModule } from './modules/lorem';
export type { MusicModule } from './modules/music';
export { Sex } from './modules/person';
export type {
  /** @deprecated Use PersonModule instead */
  PersonModule as NameModule,
  PersonModule,
  SexType,
} from './modules/person';
export type { PhoneModule } from './modules/phone';
export type { RandomModule } from './modules/random';
export type { ChemicalElement, ScienceModule, Unit } from './modules/science';
export type { StringModule } from './modules/string';
export type { SystemModule } from './modules/system';
export type { VehicleModule } from './modules/vehicle';
export type { WordModule } from './modules/word';
export { Faker };

// since we are requiring the top level of faker, load all locales by default
export const faker: Faker = new Faker({
  locales: allLocales,
});
