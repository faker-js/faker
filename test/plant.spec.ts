import { beforeEach, describe, expect, it } from 'vitest';
import { faker } from '../src';
import { seededTests } from './support/seededRuns';

const NON_SEEDED_BASED_RUN = 5;

describe('plant', () => {
  beforeEach(() => {
    faker.locale = 'en';
  });

  seededTests(faker, 'plant', (t) => {
    t.itEach(
      'plantName',
      'plantDescription',
      'plantCategory',
      'plantPotSize',
      'plantOrigin'
    );
  });

  describe(`random seeded tests for seed ${faker.seed()}`, () => {
    for (let i = 1; i <= NON_SEEDED_BASED_RUN; i++) {
      describe('name()', () => {
        it('should return a name', () => {
          const name = faker.plant.plantName();

          expect(name).toBeTruthy();
          expect(name).toBeTypeOf('string');
          expect(faker.definitions.plant?.name).toContain(name);
        });
      });

      describe('description()', () => {
        it('returns a random description', () => {
          const description = faker.plant.plantDescription();

          expect(description).toBeTruthy();
          expect(description).toBeTypeOf('string');
          expect(faker.definitions.plant?.description).toContain(description);
        });
      });

      describe('category()', () => {
        it('returns a random category', () => {
          const category = faker.plant.plantCategory();

          expect(category).toBeTruthy();
          expect(category).toBeTypeOf('string');
          expect(faker.definitions.plant?.category).toContain(category);
        });
      });

      describe('potSize()', () => {
        it('returns a random pot size', () => {
          const potSize = faker.plant.plantPotSize();

          expect(potSize).toBeTruthy();
          expect(potSize).toBeTypeOf('string');
          expect(faker.definitions.plant?.potSize).toContain(potSize);
        });
      });

      describe('origin()', () => {
        it('returns a random origin', () => {
          const origin = faker.plant.plantOrigin();

          expect(origin).toBeTruthy();
          expect(origin).toBeTypeOf('string');
        });
      });
    }
  });
});
