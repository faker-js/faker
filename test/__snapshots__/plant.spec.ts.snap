// Vitest Snapshot v1

exports[`plant > 42 > plantCategory 1`] = `"Dieffenbachia"`;

exports[`plant > 42 > plantDescription 1`] = `"The Calathea originates from the tropical rain forests of Southern America. In these rainforests they grow in the shadows of trees and other big plants, which is why they are also called “Shadow plant”."`;

exports[`plant > 42 > plantName 1`] = `"Charlotte"`;

exports[`plant > 42 > plantOrigin 1`] = `"Haiti"`;

exports[`plant > 42 > plantPotSize 1`] = `"34 cm"`;

exports[`plant > 1211 > plantCategory 1`] = `"Stromanthe"`;

exports[`plant > 1211 > plantDescription 1`] = `"The Sansevieria Fernwood Mikado is an adorable, spikey lady that owes her name ‘Mikado’ to the old-school game. She doesn’t require much care as she originally lives in the desert, so she is used to little water. She differs from other plants as she has really cool upright cylindrical leaves that end in a point at the top. Her tough tiger-striped pattern makes her special appearance complete."`;

exports[`plant > 1211 > plantName 1`] = `"Fabian"`;

exports[`plant > 1211 > plantOrigin 1`] = `"Uganda"`;

exports[`plant > 1211 > plantPotSize 1`] = `"49 cm"`;

exports[`plant > 1337 > plantCategory 1`] = `"Castano"`;

exports[`plant > 1337 > plantDescription 1`] = `"The Philodendron Gloriosum now also in this baby size. This baby is a beautiful bright green plant with, when grown up, large heart-shaped leaves. Her leaves are a beautiful green with striking yellow veins."`;

exports[`plant > 1337 > plantName 1`] = `"White Wizard"`;

exports[`plant > 1337 > plantOrigin 1`] = `"Equatorial Guinea"`;

exports[`plant > 1337 > plantPotSize 1`] = `"31 cm"`;
